### ISSUE TITLE
vpx image corruption

### ISSUE THREAD

**Issue by totaam** [Xpra-org/xpra#3187]
Could be related to #3167, #3031.
Easily reproducible by running `glxgears` with the following server settings:
```
xpra start \
    --start=glxgears \
    --encodings=jpeg,rgb,png \
    --video-encoders=vpx \
    --csc-modules=libyuv -d compress
```
(also occurs with `swscale`)

**Comment by totaam** [comment]
Bisected:
c4d8484af657656dbbe74f3f54a75ad6123794b6 is the first bad commit

And it's huge.

**Comment by stdedos** [comment]
I still see this, after the server upgrade.
Do I need to update the client as well? 😕

S: r29386 (adea604c2) / C: r29345 (9096ffbb4) with r29365 available.

**Comment by totaam** [comment]
> I still see this, after the server upgrade.

Visual corruption with vpx? Are you sure that it is with vpx? (`XPRA_PAINT_BOX=1`)

> Do I need to update the client as well?

No, the fix is server-side only.

**Comment by stdedos** [comment]
> Visual corruption with vpx? Are you sure that it is with vpx? 

I am sure it creates these logs 😕 

> https://github.com/Xpra-org/xpra/issues/3167#issuecomment-871199162
> Tons and tons of
> 
> ```
> Warning: client does not support any csc modes with ... on window ...
> ```
> 
> and
> 
> ```
>  181561 2021-06-30 11:12:37,661 using non-video fallback encoding
>  181562 2021-06-30 11:20:51,998 Error: no video pipeline options found for BGRX 24-bit at 2163x1284
>  181563 2021-06-30 11:20:51,998 Error: failed to setup a video pipeline for auto encoding with source format BGRX
>  181564 2021-06-30 11:20:51,999  all encoders: 
>  181565 2021-06-30 11:20:51,999  supported CSC modes: 
>  181566 2021-06-30 11:20:51,999  supported encoders: 
>  181567 2021-06-30 11:20:51,999  encoders CSC modes: 
> ```
> 
> these. I guess this is "your" doing?


> (`XPRA_PAINT_BOX=1`)
Can I use the shortcut to do that? Or does that have to be activated on the server?

**Comment by totaam** [comment]
> `Warning: client does not support any csc modes with`
> ...

So, no visual corruption, right?

I believe that those warnings come from the fact that you used `--video-encoders=vpx --csc-modules=libyuv` from the client side (as per https://github.com/Xpra-org/xpra/issues/3167#issuecomment-873931344) and that you have opengl disabled.  
In that case, there are no encoders + CSC + decoders combinations that can produce the RGB pixel data needed by the non-opengl renderer.

>> `(XPRA_PAINT_BOX=1)`
>> 
> Can I use the shortcut to do that? Or does that have to be activated on the server?

That's a client-side switch only.
It can be activated with the default key shortcut `#+F12:toggle_debug` (and `#` is `Control+Shift` on MS Windows I believe?)

**Comment by stdedos** [comment]
> So, no visual corruption, right?

None that I've seen, no. (I have seen some visual corruption recently, but nothing related to this ticket)

> I believe that those warnings come from the fact that you used `--video-encoders=vpx --csc-modules=libyuv` from the client side (as per [#3167 (comment)](https://github.com/Xpra-org/xpra/issues/3167#issuecomment-873931344)) and that you have opengl disabled.
> In that case, there are no encoders + CSC + decoders combinations that can produce the RGB pixel data needed by the non-opengl renderer.

So I guess "as-defined"? Weird that this helped my case these days.
I'll try to work without those flags and see the responsiveness that I'm getting.

> It can be activated with the default key shortcut `#+F12:toggle_debug` (and `#` is `Control+Shift` on MS Windows I believe?)

Yeah, that's the one I was talking about (`Alt+Shift` on Windows).

**Comment by stdedos** [comment]
`--csc-modules=help --csc-modules=libyuv` does not trigger the help text, instead connects normally

**Comment by stdedos** [comment]
Activating `swscale` now gives:

```
processing draw packet
Traceback (most recent call last):
  File "E:\xpra/xpra/client/mixins/window_manager.py", line 1399, in _draw_thread_loop
  File "E:\xpra/xpra/client/mixins/window_manager.py", line 1468, in _do_draw
  File "E:\xpra/xpra/client/client_window_base.py", line 741, in draw_region
  File "E:\xpra/xpra/client/window_backing_base.py", line 697, in draw_region
  File "E:\xpra/xpra/client/window_backing_base.py", line 605, in paint_with_video_decoder
  File "E:\xpra/xpra/client/window_backing_base.py", line 636, in do_video_paint
  File "E:\xpra/xpra/client/window_backing_base.py", line 511, in make_csc
Exception: no csc module found for wid 32 GBRP(2x20) to BGRA or BGRX or RGBA or RGBX or BGR or RGB or r210 or BGR565(2x20) in {'BGRX': {'YUV420P': [libyuv(BGRX to YUV420P), swscale(BGRX to YUV420P)], 'NV12': [libyuv(BGRX to NV12), swscale(BGRX to NV12)], 'RGB': [swscale(BGRX to RGB)], 'BGR': [swscale(BGRX to BGR)], 'XRGB': [swscale(BGRX to XRGB)], 'RGBX': [swscale(BGRX to RGBX)], 'XBGR': [swscale(BGRX to XBGR)], 'YUV422P': [swscale(BGRX to YUV422P)], 'YUV444P': [swscale(BGRX to YUV444P)], 'GBRP': [swscale(BGRX to GBRP)]}, 'RGB': {'BGR': [swscale(RGB to BGR)], 'XRGB': [swscale(RGB to XRGB)], 'BGRX': [swscale(RGB to BGRX)], 'RGBX': [swscale(RGB to RGBX)], 'XBGR': [swscale(RGB to XBGR)], 'YUV420P': [swscale(RGB to YUV420P)], 'YUV422P': [swscale(RGB to YUV422P)], 'YUV444P': [swscale(RGB to YUV444P)], 'GBRP': [swscale(RGB to GBRP)], 'NV12': [swscale(RGB to NV12)]}, 'BGR': {'RGB': [swscale(BGR to RGB)], 'XRGB': [swscale(BGR to XRGB)], 'BGRX': [swscale(BGR to BGRX)], 'RGBX': [swscale(BGR to RGBX)], 'XBGR': [swscale(BGR to XBGR)], 'YUV420P': [swscale(BGR to YUV420P)], 'YUV422P': [swscale(BGR to YUV422P)], 'YUV444P': [swscale(BGR to YUV444P)], 'GBRP': [swscale(BGR to GBRP)], 'NV12': [swscale(BGR to NV12)]}, 'XRGB': {'RGB': [swscale(XRGB to RGB)], 'BGR': [swscale(XRGB to BGR)], 'BGRX': [swscale(XRGB to BGRX)], 'RGBX': [swscale(XRGB to RGBX)], 'XBGR': [swscale(XRGB to XBGR)], 'YUV420P': [swscale(XRGB to YUV420P)], 'YUV422P': [swscale(XRGB to YUV422P)], 'YUV444P': [swscale(XRGB to YUV444P)], 'GBRP': [swscale(XRGB to GBRP)], 'NV12': [swscale(XRGB to NV12)]}, 'RGBX': {'RGB': [swscale(RGBX to RGB)], 'BGR': [swscale(RGBX to BGR)], 'XRGB': [swscale(RGBX to XRGB)], 'BGRX': [swscale(RGBX to BGRX)], 'XBGR': [swscale(RGBX to XBGR)], 'YUV420P': [swscale(RGBX to YUV420P)], 'YUV422P': [swscale(RGBX to YUV422P)], 'YUV444P': [swscale(RGBX to YUV444P)], 'GBRP': [swscale(RGBX to GBRP)], 'NV12': [swscale(RGBX to NV12)]}, 'XBGR': {'RGB': [swscale(XBGR to RGB)], 'BGR': [swscale(XBGR to BGR)], 'XRGB': [swscale(XBGR to XRGB)], 'BGRX': [swscale(XBGR to BGRX)], 'RGBX': [swscale(XBGR to RGBX)], 'YUV420P': [swscale(XBGR to YUV420P)], 'YUV422P': [swscale(XBGR to YUV422P)], 'YUV444P': [swscale(XBGR to YUV444P)], 'GBRP': [swscale(XBGR to GBRP)], 'NV12': [swscale(XBGR to NV12)]}, 'YUV420P': {'RGB': [swscale(YUV420P to RGB)], 'BGR': [swscale(YUV420P to BGR)], 'XRGB': [swscale(YUV420P to XRGB)], 'BGRX': [swscale(YUV420P to BGRX)], 'RGBX': [swscale(YUV420P to RGBX)], 'XBGR': [swscale(YUV420P to XBGR)], 'YUV422P': [swscale(YUV420P to YUV422P)], 'YUV444P': [swscale(YUV420P to YUV444P)], 'NV12': [swscale(YUV420P to NV12)]}, 'YUV422P': {'RGB': [swscale(YUV422P to RGB)], 'BGR': [swscale(YUV422P to BGR)], 'XRGB': [swscale(YUV422P to XRGB)], 'BGRX': [swscale(YUV422P to BGRX)], 'RGBX': [swscale(YUV422P to RGBX)], 'XBGR': [swscale(YUV422P to XBGR)], 'YUV420P': [swscale(YUV422P to YUV420P)], 'YUV444P': [swscale(YUV422P to YUV444P)], 'NV12': [swscale(YUV422P to NV12)]}, 'YUV444P': {'RGB': [swscale(YUV444P to RGB)], 'BGR': [swscale(YUV444P to BGR)], 'XRGB': [swscale(YUV444P to XRGB)], 'BGRX': [swscale(YUV444P to BGRX)], 'RGBX': [swscale(YUV444P to RGBX)], 'XBGR': [swscale(YUV444P to XBGR)], 'YUV420P': [swscale(YUV444P to YUV420P)], 'YUV422P': [swscale(YUV444P to YUV422P)], 'GBRP': [swscale(YUV444P to GBRP)], 'NV12': [swscale(YUV444P to NV12)]}, 'GBRP': {'RGB': [swscale(GBRP to RGB)], 'BGR': [swscale(GBRP to BGR)], 'XRGB': [swscale(GBRP to XRGB)], 'BGRX': [swscale(GBRP to BGRX)], 'RGBX': [swscale(GBRP to RGBX)], 'XBGR': [swscale(GBRP to XBGR)], 'YUV420P': [swscale(GBRP to YUV420P)], 'YUV422P': [swscale(GBRP to YUV422P)], 'YUV444P': [swscale(GBRP to YUV444P)], 'NV12': [swscale(GBRP to NV12)]}, 'NV12': {'RGB': [swscale(NV12 to RGB)], 'BGR': [swscale(NV12 to BGR)], 'XRGB': [swscale(NV12 to XRGB)], 'BGRX': [swscale(NV12 to BGRX)], 'RGBX': [swscale(NV12 to RGBX)], 'XBGR': [swscale(NV12 to XBGR)], 'YUV420P': [swscale(NV12 to YUV420P)], 'YUV422P': [swscale(NV12 to YUV422P)], 'YUV444P': [swscale(NV12 to YUV444P)], 'GBRP': [swscale(NV12 to GBRP)]}}
```
on the client ...

**Comment by stdedos** [comment]
Activating `swscale` gave couple of spinners, but not consistently.
Activating `cython` immediately after `swscale`s timeouts, gave spinners immediately upon "real" use/

**Comment by totaam** [comment]
> Activating `swscale` now gives:

Please include the full command lines of both ends. I'm lost.

**Comment by stdedos** [comment]
> Please include the full command lines of both ends. I'm lost.

Server has "no specific configuration" when started (except I think --opengl=no, if it gets propagated from `xpra start` from the client). (but I don't have it right now). 
Client has `"Xpra-Python3-x86_64_4.3-r29346\xpra_cmd" attach ssh://user@ip/2 --ssh="plink -ssh -agent" --title="@title@ on @@/@server-display@" --csc-modules=libyuv,swscale --opengl=no`

**Comment by totaam** [comment]
Thanks @stdedos!
That was very easily reproducible and this is exactly the same bug fix as f55b6575743882efb61fe394c4d70bf973d21c9d.

I clearly remember thinking that I should go through all the changes and look for the same mistake everywhere. I must have forgotten...
